service/storage_service.cc

```diff
@@ -3266,6 +3268,16 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
     } ();
 
     co_await _gossiper.wait_for_gossip_to_settle();
+
+    // This is the moment when the locator::topology has gathered information about other nodes
+    // in the cluster -- either through gossiper, or by loading it from disk -- so it's safe
+    // to start the hint managers.
+    if (start_hint_manager) {
+        co_await proxy.invoke_on_all([&gossiper] (storage_proxy& local_proxy) {
+            return local_proxy.start_hints_manager(gossiper.local().shared_from_this());
```

Comment:
You postpone start of hints_manager so that it happens after topology is available.

Doesn't the same logic apply to shutdown order? You didn't change the shutdown order. Is topology still available when hints_manager is stopped? If not, perhaps we should move shutdown of hints_manager earlier as well?

Generally, shutdown/deinitialization order should be the exact reverse of startup/initialization.

So if init of proxy is moved until after topology is available, then deinit of proxy should be moved before topology is destroyed.

Unless topology is never destroyed. Still, perhaps it's good to keep the order correspondence for sanity.

Reply:
@kbr-scylla, in the current implementation, the hint manager is stopped as part of the draining process of storage proxy. It still seems to be the right place to do that to me.

It may happen, if these changes are merged, that we start draining storage proxy before the hint managers are started – but in such a case, `db::hints::manager::stop()` will be equivalent to a no-op, so nothing should break compared to now.

Some steps are still taken, though. If you look at `storage_proxy::on_leave_cluster()`, we call the method `drain_for` of the hint manager. If the argument passed to that function happens to be the local node's IP, it will be more-or-less equivalent to stopping the hint manager – we'll drain all of the hints that have been stored so far and prevent accepting new hints. The hint manager is effectively frozen.

We could branch in that function checking if the IP we get is the local node's IP and if it is, call `stop()` rather than `drain_for()` if you think that would be clearer/saner.